third_party/move/move-vm/types/src/loaded_data/runtime_access_specifier.rs

```diff
@@ -0,0 +1,527 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Runtime representation of access control specifiers.
+//!
+//! Specifiers are represented as a list of inclusion and exclusion clauses. Each
+//! of those clauses corresponds to an `acquires A`, `reads A`, or `writes A`
+//! declaration in the language. Exclusions stem from negation, e.g. `!reads A`.
+//!
+//! Specifiers support access check via `AccessSpecifier::allows`. Moreover,
+//! access specifiers can be joined via `AccessSpecifier::join`. Joining happens
+//! when a function is entered which has access specifiers: then the current active access
+//! specifier is joined with the function's specifier.
+//!
+//! The `join` operation attempts to simplify the resulting access specifier, making
+//! access checks faster and keeping memory use low. This is only implemented for
+//! inclusions, which are fully simplified. Exclusions are accumulated.
+//! There is potential for optimization by simplifying exclusions are effectively negations,
+//! such a simplification is not trivial and may require recursive specifiers, which
+//! we like to avoid.
+
```

Comment:
Do access control specifiers exactly specify the set of resources which may be accessed, or are they an approximation?  If an approximation, which way can errors go?

If this is correct, can we specify it explicitly?  'join' operation corresponds to intersection of the sets of possibly accessed resources.

What does "specialize" mean?  I'm guessing, that any address parameters mentioned in an AccessSpecifier are made specific given the address values available at runtime.  Can we have an explicit comment about that somewhere?  Perhaps on the "specialize" method.

Reply:
Tried to reformulate and be a bit more precise, take a look again.